src/System.Memory/tests/ReadOnlySpan/Overlaps.cs

```diff
@@ -0,0 +1,990 @@
+        {
+            if (source.Length != destination.Length)
+                throw new ArgumentException();
+
+            // This loop below moves backwards, so if there is an overlap and destination starts
+            // before source then the loop will overwrite unread data without making a copy first.
+
+            if (source.Overlaps(destination, out int elementOffset) && elementOffset < 0)
+                source = source.ToArray();
+
+            for (int i = source.Length - 1; i >= 0; i--)
+                destination[i] = 2 * source[i];
+        }
+
+        [Fact]
+        public static void TestAlignedBackwards()
+        {
+            for (int i = 0; i < 14; i++)
+            {
+                int[] a = new int[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19 };
+
+                ReadOnlySpan<int> source = a.AsReadOnlySpan().Slice(7, 5);
+
+                Span<int> expected = new int[a.Length].AsSpan().Slice(i, 5);
+                Span<int> actual = a.AsSpan().Slice(i, 5);
+
+                DoubleEachElementBackwards(source, expected);
+                DoubleEachElementBackwards(source, actual);
+
+                Assert.Equal(expected.ToArray(), actual.ToArray());
+            }
+        }
+
+        [Fact]
+        public static void TestUnalignedBackwards()
+        {
+            Assert.Throws<ArgumentException>(() =>
+            {
+                int[] a = new int[] { 1, 2, 3, 4, 5, 6 };
+
+                ReadOnlySpan<int> source = a.AsReadOnlySpan().AsBytes()
+                    .Slice(2, 5 * sizeof(int))
+                    .NonPortableCast<byte, int>();
+
+                Span<int> actual = a.AsSpan().Slice(0, 5);
+
+                DoubleEachElementBackwards(source, actual);
+            });
+
+            Assert.Throws<ArgumentException>(() =>
+            {
+                int[] a = new int[] { 1, 2, 3, 4, 5, 6 };
+
+                ReadOnlySpan<int> source = a.AsReadOnlySpan().AsBytes()
+                    .Slice(2, 5 * sizeof(int))
+                    .NonPortableCast<byte, int>();
+
+                Span<int> actual = a.AsSpan().Slice(1, 5);
+
+                DoubleEachElementBackwards(source, actual);
+            });
+        }
+
+        [Fact]
+        public static void SizeOf1Overlaps()
+        {
+            byte[] a = new byte[16];
+
+            Assert.True(a.AsReadOnlySpan().Slice(0, 12).Overlaps(a.AsReadOnlySpan().Slice(8, 8), out int elementOffset));
+            Assert.Equal(8, elementOffset);
+        }
+
+        [Fact]
+        public static void SizeOf16Overlaps()
+        {
+            Guid[] a = new Guid[16];
+
+            Assert.True(a.AsReadOnlySpan().Slice(0, 12).Overlaps(a.AsReadOnlySpan().Slice(8, 8), out int elementOffset));
+            Assert.Equal(8, elementOffset);
+        }
+
+        //
+        // The following tests were all generated with by this (otherwise unused) method:
```

Comment:
nit: fix comment "...generated with by this..."

Reply:
Fixed.